### PR TITLE
Error if any invalid keys are defined when inheriting a dependency

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

PRs in this RFC:
- Base Implementation
  - #10497
  - #10517 
  - #10538 
  - #10548 
- New Fields 
  - #10563
  - #10565 
- Update inheritance source 
  - #10564
- Optimizations
  - #10568 
- Documentation changes
  - #10584
- `cargo-add` support
  - #10585
  - #1
  - #3

Changes:
- Added checks for invalid keys (default-features, registry, branch, tag, rev) when inheriting a dependency from a workspace

Remaining implementation work for the RFC
- `cargo-add` support, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1075544790)
- Optimizations as needed

r? @epage

### CODE REVIEW

**Review by epage — CHANGES_REQUESTED** [review]
We will have everything pre-parsed by the regular parser, so we shouldn't need to error check this.

In my re-wording, I think some context was lost.  We need to error if the user sets any flags on the command-line that are incompatible with `workspace = true`.

**Comment by Muscraft** [comment]
So the error should be during the parsing of the CLI flags?

**Comment by epage** [comment]
> So the error should be during the parsing of the CLI flags?

It has to happen where we have access to the CLI flags and we know its a workspace dependency.  Note that we pass in a struct to the ops that basically represents the CLI, so we know the details.

Keep in mind that right now we are just loading an existing workspace entry.  We'll soon also take the name and look up to find if its a workspace dependency.

**Comment by Muscraft** [comment]
Would it make sense to wait until after adding in where we take the name and look up to find if it's a workspace dependency, to do this?

**Comment by epage** [comment]
We can implement the PRs in any order so long as we keep the work for both in mind though doing the other PR first will make this one more obvious for what to do

**Comment by Muscraft** [comment]
I agree that one of the others would help make this slightly more clear. I'll do one of those and then come back to this.

File: tests/testsuite/cargo_add.rs [hunk 1]
```diff
@@ -1026,6 +1026,28 @@ fn invalid_git_external() {
     );
 }
 
+#[cargo_test]
+fn invalid_key_inherit_dependency() {
```
Comment (epage) [inline tests/testsuite/cargo_add.rs:1030]: What about the case where the dependency already exists and we use one of these keys?

File: src/cargo/ops/cargo_add/mod.rs [hunk 2]
```diff
@@ -284,6 +284,22 @@ fn resolve_dependency(
         // Checking for a workspace dependency happens first since a member could be specified
         // in the workspace dependencies table as a dependency
         if let Some(_dep) = find_workspace_dep(dependency.toml_key(), ws.root_manifest()).ok() {
+            let err_msg = "cannot be defined when inheriting a workspace dependency";
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs]: What if we rephrase this as
```
 "cannot override workspace dependency with `{}`, either change `workspace.dependencies.{}.{}` or define the dependency exclusively in the package's manifest"
```
- I feel like "defined" is confusing language in this context
- I feel like talking in terms of "overriding" is more specific and clear than generically talking about "inheriting"
- This provides a remediation

File: src/cargo/ops/cargo_add/mod.rs [hunk 3]
```diff
@@ -284,6 +284,22 @@ fn resolve_dependency(
...
+            }
+            if arg.branch.is_some() {
+                anyhow::bail!("`branch` {}", err_msg)
+            }
+            if arg.tag.is_some() {
+                anyhow::bail!("`tag` {}", err_msg)
+            }
+            if arg.rev.is_some() {
+                anyhow::bail!("`rev` {}", err_msg)
+            }
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs]: > . This means you cannot define workspace with keys like version, registry, registry-index, path, git, branch, tag, rev, or package.

What about checking for
- path
- git
- version
- package

Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: The only package that is used here is to tell cargo add what package to add the dependency too. The others should be checked before this since they all would set a source from what I can tell

File: src/cargo/ops/cargo_add/mod.rs (hunk 2 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: Does this override anything? To me, this only runs when it is not overriding something since it would've been found by `get_existing_dependency`. As for changing it I agree that we should change it to something else

File: src/cargo/ops/cargo_add/mod.rs (hunk 3 above)
Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs]: > The only package that is used here is to tell cargo add what package to add the dependency too.

Try a test case where you `cargo add --rename foo foo-alt` where `foo` is a workspace dependency and `foo-alt` is a workspace member.

> The others should be checked before this since they all would set a source from what I can tell

Thats right, we either do or will cover this with testing switching sources.  Probably good to add a comment about it.

File: src/cargo/ops/cargo_add/mod.rs (hunk 2 above)
Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs]: I guess there are two different kinds of overriding to talk about.  This isn't writing over an existing fields value with a different one but it is instead overriding the field it inherits from `workspace.dependencies`.

File: tests/testsuite/cargo_add.rs (hunk 1 above)
Comment (Muscraft) [reply tests/testsuite/cargo_add.rs:1030]: I can add a test for this

File: src/cargo/ops/cargo_add/mod.rs (hunk 2 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: Yeah, that makes sense. I'll change it to what you said

File: src/cargo/ops/cargo_add/mod.rs (hunk 3 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: > Try a test case where you `cargo add --rename foo foo-alt` where `foo` is a workspace dependency and `foo-alt` is a workspace member.

So `package` is `rename`? or at least that's the way it looks to me

> Thats right, we either do or will cover this with testing switching sources. Probably good to add a comment about it.

I'll add a comment about this

Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: I also do not know if we need `branch`, `tag`, or `rev`, since they can't be defined on their own

Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs]: I was thinking similar about branch, tag, and rev.

File: src/cargo/ops/cargo_add/mod.rs [hunk 4]
```diff
@@ -328,6 +329,38 @@ fn resolve_dependency(
...
+/// `version`  should all bee checked before this is called. `rename` is checked
+/// for as it turns into `package`
+fn check_invalid_ws_keys(toml_key: &str, arg: &DepOp) -> CargoResult<()> {
+    fn err_msg(toml_key: &str, key: &str) -> String {
+        format!(
+            "cannot override workspace dependency with `{}`, \
+            either change `workspace.dependencies.{}.{}` \
+            or define the dependency exclusively in the package's manifest",
+            key, toml_key, key
+        )
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs]: nit: this might read better if using the new format macro feature
```
            "cannot override workspace dependency with `{field}`, \
            either change `workspace.dependencies.{toml_key}.{field}` \
            or define the dependency exclusively in the package's manifest",
```

File: tests/snapshots/add/invalid_key_overwrite_inherit_dependency.in/primary/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+cargo-features = ["workspace-inheritance"]
+
+[package]
+name = "bar"
+version = "0.0.0"
+
+foo.workspace = true
```
Comment (epage) [inline tests/snapshots/add/invalid_key_overwrite_inherit_dependency.in/primary/Cargo.toml]: Is this missing a `[dependencies]`?

File: src/cargo/ops/cargo_add/mod.rs [hunk 6]
```diff
@@ -328,6 +329,38 @@ fn resolve_dependency(
...
+
+    if arg.default_features.is_some() {
+        anyhow::bail!("{}", err_msg(toml_key, "default-features"))
+    }
+    if arg.registry.is_some() {
+        anyhow::bail!("{}", err_msg(toml_key, "registry"))
+    }
+    // rename is `package`
+    if arg.rename.is_some() {
+        anyhow::bail!("{}", err_msg(toml_key, "rename"))
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs]: This is a case where the argument and field diverge (`--rename` to cause `package` to be set)

In general, I feel like the error message will be easier to read with arguments
```
error: cannot override workspace dependency with `--rename`, either change `workspace.dependencies.foo.package` or define the dependency exclusively in the package's manifest
```

File: tests/testsuite/cargo_add.rs [hunk 7]
```diff
@@ -1026,6 +1026,74 @@ fn invalid_git_external() {
...
+fn invalid_key_overwrite_inherit_dependency() {
+    let project =
+        Project::from_template("tests/snapshots/add/invalid_key_overwrite_inherit_dependency.in");
+    let project_root = project.root();
+    let cwd = &project_root;
+
+    snapbox::cmd::Command::cargo()
+        .masquerade_as_nightly_cargo()
+        .arg("add")
+        .args(["foo", "--registry", "registry", "-p", "bar"])
```
Comment (epage) [inline tests/testsuite/cargo_add.rs]: Is there a reason this test is using `--registry`?  I think it can obscure what you are trying to test by changing more variables.

File: src/cargo/ops/cargo_add/mod.rs (hunk 6 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]:
```rust
fn err_msg(toml_key: &str, flag: &str, field: &str) -> String {
        format!(
            "cannot override workspace dependency with `{flag}`, \
            either change `workspace.dependencies.{toml_key}.{field}` \
            or define the dependency exclusively in the package's manifest"
        )
    }

    if arg.default_features.is_some() {
        anyhow::bail!("{}", err_msg(toml_key, "--default-features", "default-features"))
    }
    if arg.registry.is_some() {
        anyhow::bail!("{}", err_msg(toml_key, "--registry", "registry"))
    }
    // rename is `package`
    if arg.rename.is_some() {
        anyhow::bail!("{}", err_msg(toml_key, "--rename", "package"))
    }
```

File: tests/snapshots/add/invalid_key_overwrite_inherit_dependency.in/primary/Cargo.toml (hunk 5 above)
Comment (Muscraft) [reply tests/snapshots/add/invalid_key_overwrite_inherit_dependency.in/primary/Cargo.toml]: It looks like it is

File: src/cargo/ops/cargo_add/mod.rs (hunk 4 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: I'll change it to this

File: tests/testsuite/cargo_add.rs (hunk 7 above)
Comment (Muscraft) [reply tests/testsuite/cargo_add.rs]: There is no specific reason that used `--registry` here, do you think I should change it to `--default-features`?

Comment (epage) [reply tests/testsuite/cargo_add.rs]: Yes, please

File: src/cargo/ops/cargo_add/mod.rs [hunk 8]
```diff
@@ -272,6 +272,9 @@ fn resolve_dependency(
                 // Overwrite with `crate_spec`
                 old_dep.source = selected_dep.source;
             }
+            if let Some(Source::Workspace(_)) = old_dep.source() {
+                check_invalid_ws_keys(old_dep.toml_key(), arg)?;
+            }
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs]: What if we combined the two and moved this to the end of the function?
- Check the source on `dependency` instead of `old_dep`
- Condition the check on the source being a Workspace

**Review by epage — COMMENTED** [review]
Once/if the checks get consolidated

**Review by epage — APPROVED** [review]
Ok, this time I remembered to actually approve

File: src/cargo/ops/cargo_add/mod.rs (hunk 8 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: That works for me. Does moving it to after `if dependency.source().is_none()`, instead of the end of the function work?

Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs]: It doesn't have to be literally at the end of the function but among the clauses at the end of the function.

Stylistically, make sure it has newlines before and after since its not logically coupled to any of the blocks.

Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs]: That's how I have it! I'll push the changes and merge it then!
